File: android/BinaryTreeTraversal/app/src/main/java/edu/uoregon/parsab/binarytreetraversal/BST.java
package edu.uoregon.parsab.binarytreetraversal;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Stack;

public class BST<E extends Comparable<E>> implements Parcelable {
    private Node<E> root;

    public BST(){
        root = null;
    }

    protected BST(Parcel in) {
        root = (Node<E>) in.readSerializable();
    }

    public static final Creator<BST> CREATOR = new Creator<BST>() {
        @Override
        public BST createFromParcel(Parcel in) {
            return new BST(in);
        }

        @Override
        public BST[] newArray(int size) {
            return new BST[size];
        }
    };

    public Node<E> getRoot(){
        return root;
    }

    private Node<E> insertNode(Node<E> current, Node<E> n){
        if(current == null)
            current = n;
        else if(current.getData().compareTo(n.getData())< 0) {
            current.setRightChild(insertNode(current.getRightChild(), n));
            current.getRightChild().setParent(current);
        }
        else if(current.getData().compareTo(n.getData())>0) {
            current.setLeftChild(insertNode(current.getLeftChild(), n));
            current.getLeftChild().setParent(current);
        }
        return current;
    }
    public void insert(E data){
        Node<E> node = new Node<E>(data);
        root = insertNode(root, node);
    }

    public Node<E> find(E data){
        boolean done = false;
        Node<E> temp = root;


        while(!done){
            if (temp == null){
                return null;
            }
            if(temp.getData().compareTo(data) == 0){
                done = true;
            } else if (temp.getData().compareTo(data) > 0){
                temp = temp.getLeftChild();
            } else if (temp.getData().compareTo(data) < 0){
                temp = temp.getRightChild();
            }
        }
        return temp;
    }
    private Node<E> findDeleteNode(Node<E> current, E val) {
        if(current == null) {	// If we get to a leaf or null, it means we didn't find the node.
            System.out.println("Node not found.");
        }
        else if(current.getData().compareTo(val) == 0){	// Node has been found. Perform delete operation.
            return deleteCurrentNode(current);
        }
        else if(current.getData().compareTo(val) > 0){	//Traversing to the left
            current.setLeftChild(findDeleteNode(current.getLeftChild(), val));
            if(current.getLeftChild()!= null)
                current.getLeftChild().setParent(current);
        }
        else if(current.getData().compareTo(val) < 0){	//Traversing to the right
            current.setRightChild(findDeleteNode(current.getRightChild(), val));
            if(current.getRightChild()!= null)
                current.getRightChild().setParent(current);
        }
        return current;
    }
    private int IdentifyCase(Node<E> current) {
        if(current.getLeftChild() == null && current.getRightChild() == null){
            return 1;
        }
        else if(current.getLeftChild() != null && current.getRightChild() != null){
            return 3;
        }
        return 2;
    }
    private E findLowestInRightSubTree(Node<E> current) {
        E val = current.getData();
        while(current.getLeftChild() != null){
            current = current.getLeftChild();
            val = current.getData();
        }
        return val;
    }
    private Node<E> deleteCase2(Node<E> current) {
        if(current.getLeftChild() == null){
            return current.getRightChild();
        }
        return current.getLeftChild();
    }
    private Node<E> deleteCase3(Node<E> current) {
        E replaceVal = findLowestInRightSubTree(current.getRightChild());	// Call a method to find the value of the replacement.
        current.setData(replaceVal);								// Set the data of the current node to the replacement value.
        current.setRightChild(findDeleteNode(current.getRightChild(), replaceVal));// Call delete on the right subtree, with the replacement value the one being deleted. Assign whatever is returned to the right subtree.
        if(current.getRightChild()!= null)
            current.getRightChild().setParent(current);
        return current;
    }
    private Node<E> deleteCurrentNode(Node<E> current) {
        int icase = IdentifyCase(current);
        if(icase == 1){ //no children
            return null;
        }
        else if(icase == 2){ // one child
            return deleteCase2(current);
        }
        else if(icase == 3){ //two children
            return deleteCase3(current);
        }
        return null;
    }
    public void delete(E data) {
        root = findDeleteNode(root, data);
    }


    // in-order traversal iterative
    public String inOrder(Node<E> top){
        String traversal = "";
        if(top != null){
            Stack<Node<E>> stack = new Stack<Node<E>>();
            Node<E> node = top;
            while(node!=null){
                stack.push(node);
                node = node.getLeftChild();
            }
            while (stack.size()>0){
                node = stack.pop();
                traversal += node.getData().toString()+"  ";
                if(node.getRightChild()!=null){
                    node = node.getRightChild();
                    while(node!=null){
                        stack.push(node);
                        node = node.getLeftChild();
                    }
                }
            }
            return traversal;
        }
        return null;
    }

    //post-order traversal iterative
    public String postOrder(Node<E> top)
    {
        // Create two stacks
        Stack<Node<E>> s1 = new Stack<Node<E>>();
        Stack<Node<E>> s2 = new Stack<Node<E>>();
        String traversal = "";

        if (top != null){
            // push root to first stack
            s1.push(top);

            // Run while first stack is not empty
            while (!s1.isEmpty())
            {
                // Pop an item from s1 and push it to s2
                Node<E> temp = s1.pop();
                s2.push(temp);

                // Push left and right children of
                // removed item to s1
                if (temp.getLeftChild() != null)
                    s1.push(temp.getLeftChild());
                if (temp.getRightChild() != null)
                    s1.push(temp.getRightChild());
            }

            // Print all elements of second stack
            while (!s2.isEmpty())
            {
                Node<E> temp = s2.pop();
                traversal+= temp.getData().toString()+"  ";
            }
            return traversal;
        }
        return null;
    }

    //pre-order traversal iterative
    public String preOrder(Node<E> top) {

        if (top != null) {

            Stack<Node<E>> nodeStack = new Stack<>();
            nodeStack.push(root);
            String traversal = "";

            while (nodeStack.empty() == false) {

                Node<E> mynode = nodeStack.peek();
                traversal += mynode.getData().toString()+ "  ";
                nodeStack.pop();

                if (mynode.getRightChild() != null) {
                    nodeStack.push(mynode.getRightChild());
                }
                if (mynode.getLeftChild() != null) {
                    nodeStack.push(mynode.getLeftChild());
                }
            }
            return traversal;
        }
        return null;
    }

    public Node<E> getMin(Node<E> top){
        boolean done = false;
        Node<E> temp = top;
        while(!done) {
            if (temp.getLeftChild() == null) {
                done = true;
            } else {
                temp = temp.getLeftChild();
            }
        }
        return temp;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeSerializable(root);
    }
}

File: iOS/finalProject/Tree Traversal/Source/Model/BST.swift
//
//  BST.swift
//  Tree Traversal
//
//  Created by Parsa Bagheri on 11/21/17.
//  Copyright © 2017 Parsa Bagheri. All rights reserved.
//

class BST{
    private var root: Node?;

//################################################################################

    init(){
        root = Node(data: nil)
    }
    
    func getRoot() -> Node?{
        return root;
    }
    
//################################################################################
   // Insert Method
    
    private func insertNode( current cur: Node?, to_be_added n: Node) -> Node?{
        var current = cur
        if n.getData() == nil{
            return current
        }
        if current?.getData() == nil{
            current = n;
        }
        else if (current?.getData())! < (n.getData())!{
            current?.setRightChild(right_child: (insertNode(current: current?.getRightChild(), to_be_added: n))!)
            current?.getRightChild()?.setParent(parent: current)
        }
        else if (current?.getData())! > (n.getData())!{
            current?.setLeftChild(left_child: (insertNode(current: current?.getLeftChild(), to_be_added: n))!)
            current?.getLeftChild()?.setParent(parent: current)
        }
        return current
    }
    
    func insert(data: Float64?){
        let node: Node = Node(data: data)
        root = insertNode(current: root, to_be_added: node)
    }

//################################################################################
    //Find Method
    
    private func find(data: Float64?) -> Node?{
        var done: Bool = false
        var temp: Node? = root
        
        while done != true{
            if temp == nil{
                return nil;
            }
            if (temp?.getData())! == data!{
                done = true
            }
            else if (temp?.getData())! > data!{
                temp = temp?.getLeftChild()
            }
            else if (temp?.getData())! < data!{
                temp = temp?.getRightChild()
            }
        }
        return temp
    }
    
//################################################################################
    //Delete Method
    
    private func findDeleteNode(current: Node?, data: Float64) -> Node?{
        if current == nil{
            return nil
        }
        else if current?.getData() == data{
            return deleteCurrentNode(current: current!)
        }
        else if (current?.getData())! > data{
            current?.setLeftChild(left_child: findDeleteNode(current: current?.getLeftChild(), data: data))
            if current?.getLeftChild() != nil{
                current?.getLeftChild()?.setParent(parent: current!)
            }
        }
        else if (current?.getData())! < data{
            current?.setRightChild(right_child: findDeleteNode(current: current?.getRightChild(), data: data))
            if current?.getRightChild() != nil{
                current?.getRightChild()?.setParent(parent: current!)
            }
        }
        return current
    }
    
    private func identifyCase(current: Node?) -> Float64{
        if (current?.getLeftChild() == nil && current?.getRightChild() == nil){
            return 1
        }
        else if(current?.getLeftChild() != nil && current?.getRightChild() != nil){
            return 3
        }
        return 2
    }
    
    private func rightSubtreeLowest(current: Node) -> Float64{
        var cur: Node? = current
        var data: Float64 = (cur?.getData())!
        while cur?.getLeftChild() != nil{
            cur = cur?.getLeftChild()
            data = (cur?.getData())!
        }
        return data
    }
    
    private func deleteCase2(currnet: Node?) -> Node?{
        if currnet?.getLeftChild() == nil {
            return currnet?.getRightChild()
        }
        return currnet?.getLeftChild()
    }
    
    private func deleteCase3(current: Node) -> Node?{
        let replace_data: Float64 = rightSubtreeLowest(current: current.getRightChild()!)
        current.setData(data: replace_data)
        current.setRightChild(right_child: findDeleteNode(current: current.getRightChild(), data: replace_data))
        if current.getRightChild() != nil{
            current.getRightChild()?.setParent(parent: current)
        }
        return current
    }
    
    private func deleteCurrentNode(current: Node) -> Node?{
        let CASE: Float64 = identifyCase(current: current)
        if CASE == 1{
            return nil;
        }
        else if CASE == 2{
            return deleteCase2(currnet: current)
        }
        else if CASE == 3{
            return deleteCase3(current: current)
        }
        return nil
    }
    
    func delete (data: Float64?){
        if data != nil{
            root = findDeleteNode(current: root, data: data!)
        }
    }

//################################################################################
    // Traverse Method
    
    private func inOrder(top t: Node?) -> String{
        var traversal: String = ""
        if t?.getData() != nil{
            var stack: [Node?] = []
            var node: Node? = t
            while node != nil {
                stack.append(node)
                node = node?.getLeftChild()
            }
            while stack.count > 0 {
                node = stack.popLast() as? Node
                traversal = traversal + String(describing: (node?.getData())!) + "  "
                if node?.getRightChild() != nil{
                    node = node?.getRightChild()
                    while(node != nil){
                        stack.append(node)
                        node = node?.getLeftChild()
                    }
                }
            }
        }
        return traversal
    }
    
    private func preOrder(top t: Node?) -> String {
        var traversal: String = ""
        if t?.getData() != nil {
            var stack: [Node?] = []
            stack.append(root)
            
            while (stack.count > 0){
                let node: Node = (stack.last as? Node)!
                traversal = traversal + String(describing: (node.getData())!) + "  "
                stack.removeLast()
                if node.getRightChild() != nil {
                    stack.append(node.getRightChild())
                }
                if node.getLeftChild() != nil {
                    stack.append(node.getLeftChild())
                }
            }
        }
        return traversal
    }
    
    private func postOrder(top t: Node?) -> String {
        var stack1: [Node?] = []
        var stack2: [Node?] = []
        var traversal: String = ""
        if t?.getData() != nil {
            stack1.append(t)
            while stack1.count > 0 {
                let temp: Node = (stack1.popLast() as? Node)!
                stack2.append(temp)
                if temp.getLeftChild() != nil {
                    stack1.append(temp.getLeftChild())
                }
                if temp.getRightChild() != nil {
                    stack1.append(temp.getRightChild())
                }
            }
            while stack2.count > 0 {
                let temp: Node = (stack2.popLast() as? Node)!
                traversal = traversal + String(describing: (temp.getData())!) + "  "
            }
        }
        return traversal
    }
    
    func traverse(type: String, top t: Node?) -> String? {
        if type == "post-order"{
            return postOrder(top: t)
        }
        else if type == "pre-order"{
            return preOrder(top: t)
        }
        else{
            return inOrder(top: t)
        }
    }
}
//################################################################################
